### ISSUE TITLE
Suggestion for better pipelines with Luigi

### ISSUE THREAD

**Issue by perone** [neuropoly/spinalcordtoolbox#2169]
I saw that we often use bash scripts to build pipelines and we also often strugle with monitoring these pipelines to see if everything went well, to execute them in parallel, etc. I experimented with [Luigi](https://github.com/spotify/luigi) and it really seems to fit very well most of the use cases. Here is some example code and how to install and run it on Luigi.

### Installing Luigi
To install, just use `pip`:
```
pip install luigi
```

### Creating a simple SC segmentation task
Save this file as `sct_tasks.py`:
```python
import os
import sh
from pathlib import Path

import luigi

# No truncation for error messages
sh.ErrorReturnCode.truncate_cap = 2**64


class SCT(object):
    def __init__(self, sct_dir=None):
        if sct_dir is None:
            envdir = os.getenv("SCT_DIR")
            if envdir is None:
                raise RuntimeError("You need a proper SCT installation.")
            self.sct_root = Path(envdir)
        else:
            self.sct_root = Path(sct_dir)
        self._populate()

    def _populate(self):
        for fbin in (self.sct_root / "bin").glob("sct_*"):
            cmd = sh.Command(str(fbin.absolute()))
            setattr(self, fbin.name, cmd)

class SCTDeepSegSC(luigi.Task):
    input_file = luigi.Parameter()
    contrast = luigi.Parameter()

    def run(self):
        sct = SCT()
        output = sct.sct_deepseg_sc(i=self.input_file, c=self.contrast)

    def output(self):
        fname = Path(self.input_file)
        fname = fname.stem.split(".")[0]
        return luigi.LocalTarget(f"{fname}_seg.nii.gz")
```

### Using Luigi to execute the task
Execute the Luigi server on a terminal:
```
luigid --port=8082
```

And on another terminal, call the task:
```
python sct_tasks.py 
SCTDeepSegSC --input-file t2.nii.gz --contrast t2
```

And that's it, Luigi will execute the task.
This doesn't show much of the benefits, however, the great benefit of Luigi is that it is compositional. So we can build a task called `SCTSegmentSCGM` that will segment the SC and the GM **in parallel**, and we can create another task called `SCTSegmentSCGMDirectory` that will segment all files on a directory using this task in parallel as well (the level of paralellism can be controled too), the full example is described below:

```python
import time
import os
import sh
from pathlib import Path

import luigi

# No truncation for error messages
sh.ErrorReturnCode.truncate_cap = 2**64


class SCT(object):
    def __init__(self, sct_dir=None):
        if sct_dir is None:
            envdir = os.getenv("SCT_DIR")
            if envdir is None:
                raise RuntimeError("You need a proper SCT installation.")
            self.sct_root = Path(envdir)
        else:
            self.sct_root = Path(sct_dir)
        self._populate()

    def _populate(self):
        for fbin in (self.sct_root / "bin").glob("sct_*"):
            cmd = sh.Command(str(fbin.absolute()))
            setattr(self, fbin.name, cmd)


class SCTDeepSegSC(luigi.Task):
    input_file = luigi.Parameter()
    contrast = luigi.Parameter()

    def run(self):
        sct = SCT()
        output = sct.sct_deepseg_sc(i=self.input_file, c=self.contrast)

    def output(self):
        fname = Path(self.input_file)
        fname = fname.stem.split(".")[0]
        return luigi.LocalTarget(f"{fname}_seg.nii.gz")


class SCTDeepSegGM(luigi.Task):
    input_file = luigi.Parameter()

    def run(self):
        sct = SCT()
        output = sct.sct_deepseg_gm(i=self.input_file,
                                    o=self.output().path)

    def output(self):
        fname = Path(self.input_file)
        fname = fname.stem.split(".")[0]
        return luigi.LocalTarget(f"{fname}_gm_seg.nii.gz")


class SCTSegmentSCGM(luigi.Task):
    input_file = luigi.Parameter()
    contrast = luigi.Parameter()

    def requires(self):
        return [
            SCTDeepSegSC(input_file=self.input_file, contrast=self.contrast),
            SCTDeepSegGM(input_file=self.input_file)
        ]


class SCTSegmentSCGMDirectory(luigi.Task):
    dirpath = luigi.Parameter()

    def requires(self):
        directory = Path(self.dirpath)
        return [SCTSegmentSCGM(input_file=f, contrast="t2")
                for f in directory.glob("*.nii.gz")]


if __name__ == "__main__":
    luigi.run()
```

With this code saved on the `sct_tasks.py`, we can now execute:

```
python sct_tasks.py SCTSegmentSCGMDirectory --dirpath test/
```

and it will iterate over all files of the folder `test/` and execute the tasks in parallel. During the execution, you can visualize everything as well by accessing the UI on `http://127.0.0.1:8082`:

![image](https://user-images.githubusercontent.com/412328/53218953-2fcdfb00-362c-11e9-8345-d65c2b089c33.png)

![image](https://user-images.githubusercontent.com/412328/53219011-660b7a80-362c-11e9-9592-a11f55551918.png)

### Errors on Luigi
If an error happen during the pipeline execution, Luigi will show this in the UI:
![image](https://user-images.githubusercontent.com/412328/53219046-86d3d000-362c-11e9-9734-a5729e5a5edc.png)
So you can click on the red button and get information about the error that happened:
![image](https://user-images.githubusercontent.com/412328/53219062-92bf9200-362c-11e9-86ae-aeca4cdc8751.png)

Another interesting thing, is that you can also add an email support into Luigi, so it will warn you that something has failed or not. There is also a way to [notify on slack](https://github.com/bonzanini/luigi-slack) when pipelines finish or end up in errors.

Anyway, it seems to be simple enough to write the tasks and once you wrote them, you can just reuse them to compose other more complex tasks or even reuse them as is. Maybe moving this code I showed to a folder inside SCT would be nice to start a small repository of tasks that can be reused.

**Comment by jcohenadad** [comment]
Very nice proof-of-concept demonstration, thank you very much @perone!!!  We should explore its potential internally for multi-subject tasks, and then think of intuitive strategies for making it easy for external users to create pipelines. I anticipate this might be a bit complicated for most users, but I definitely see the advantages.

**Comment by zougloub** [comment]
Relates to #1336 (a lot).

I don't know Luigi, is it easy to declare operations that check for inputs/parameters/tool signatures, so it's possible to bypass unnecessary work when resuming/re-running pipelines ?

Since there are many schedulers (you were pointing at Dask in #1336, I used waf as a PoC, there is also [nipype](https://nipype.readthedocs.io/en/latest/) used in neuro-imaging), maybe a way to get an SCT processing node intermediate representation (maybe coming from an API documentation/introspection scheme) that can be converted to any brand of scheduler would be nice, it would avoid writing bindings for each one at some point. But a good starting point might be to just use one of them.

**Comment by jcohenadad** [comment]
Looks like there is now a trends for using NextFlow in the neuroimaging community. Closing, as this is a bit out of scope
